File: living_world/rules/heat.py
"""Tile heat scoring — pure-rule heuristic for "where is something likely
to happen right now?"

Used by the emergent-event LLM to pick which 1-2 tiles to spend its budget
on each tick. The scoring itself is deterministic and cheap (no LLM, no
embeddings) — just counts agents, historical figures, strong pairwise
affinity, and recent event density.

This file lives in `rules/` because the question "which tile is hot?" is
a pure function of world state. The downstream LLM call (in
`agents/emergent.py`) is what's intelligent.
"""

from __future__ import annotations

from living_world.core.tile import Tile
from living_world.core.world import World


def score_tile_heat(tile: Tile, world: World, *, recent_ticks: int = 3) -> float:
    """Higher = more likely something dramatic should happen here now.

    Signal weights (deliberately simple, tweak with care):
        +1.0 per resident
        +2.0 per historical figure
        +1.5 per pairwise affinity ≥ 40 (friend OR enemy — both = drama)
        +0.3 per event in this tile in the last `recent_ticks` ticks

    Returns 0.0 if fewer than 2 residents — emergent events need
    interaction, a single agent in an empty room is not interesting.
    """
    residents = world.agents_in_tile(tile.tile_id)
    if len(residents) < 2:
        return 0.0
    hf_count = sum(1 for a in residents if a.is_historical_figure)
    strong_pairs = 0
    for i, a in enumerate(residents):
        for b in residents[i + 1 :]:
            if abs(a.get_affinity(b.agent_id)) >= 40:
                strong_pairs += 1
    since = max(1, world.current_tick - recent_ticks)
    recent = sum(1 for e in world.events_since(since) if e.tile_id == tile.tile_id)
    return len(residents) + 2 * hf_count + 1.5 * strong_pairs + 0.3 * recent


def hot_tiles(world: World, *, limit: int = 2, min_heat: float = 3.5) -> list[Tile]:
    """Return the top-K tiles where something dramatic is likeliest right now."""
    scored: list[tuple[float, Tile]] = []
    for tile in world.all_tiles():
        h = score_tile_heat(tile, world)
        if h >= min_heat:
            scored.append((h, tile))
    scored.sort(key=lambda x: -x[0])
    return [t for _, t in scored[:limit]]

File: packages/sim-core/src/heat.ts
/**
 * Tile heat scoring — TypeScript port of `living_world.rules.heat`.
 *
 * Why port: the heat heuristic decides which tiles get the LLM emergent
 * proposer's budget. Today this runs in Python and the result is opaque
 * to the dashboard. Porting lets the dashboard pre-compute and DISPLAY
 * the heat ranking ("Bright Lab is hot today, expect drama") without
 * waiting for an emergent event to land.
 *
 * Pure data-in / data-out: caller pre-aggregates residents + recent
 * tile events and passes them in. The function never reaches into a
 * `World`, matching the rest of sim-core's portability rule.
 */

// ── DTOs ────────────────────────────────────────────────────────────────

export interface AgentForHeat {
  agentId: string;
  isHistoricalFigure: boolean;
  /** Map of (otherAgentId → affinity). Read-only; only |aff|≥40 entries
   *  matter for the strong-pair signal, callers may pre-filter. */
  affinityTo: Readonly<Record<string, number>>;
}

export interface TileForHeat {
  tileId: string;
  residents: ReadonlyArray<AgentForHeat>;
  /** Count of events recorded in this tile within the lookback window. */
  recentEventCount: number;
}

// ── Scoring ─────────────────────────────────────────────────────────────

/**
 * Heat score for one tile. Higher = more likely something dramatic
 * should happen there now.
 *
 * Signal weights (matching Python verbatim):
 *   +1.0 per resident
 *   +2.0 per historical figure
 *   +1.5 per pairwise |affinity| ≥ 40 (friend OR enemy — both = drama)
 *   +0.3 per event in this tile in the lookback window
 *
 * Returns 0.0 if fewer than 2 residents — emergent events need
 * interaction; a single agent in an empty room is not interesting.
 */
export function scoreTileHeat(tile: TileForHeat): number {
  const residents = tile.residents;
  if (residents.length < 2) return 0.0;
  let hfCount = 0;
  for (const a of residents) if (a.isHistoricalFigure) hfCount++;

  let strongPairs = 0;
  for (let i = 0; i < residents.length; i++) {
    const a = residents[i]!;
    for (let j = i + 1; j < residents.length; j++) {
      const b = residents[j]!;
      // a's recorded affinity toward b OR b's toward a — match Python
      // which uses Agent.get_affinity (one-sided lookup; we mirror that).
      const aff = a.affinityTo[b.agentId] ?? 0;
      if (Math.abs(aff) >= 40) strongPairs++;
    }
  }

  return (
    residents.length +
    2 * hfCount +
    1.5 * strongPairs +
    0.3 * tile.recentEventCount
  );
}

export interface HeatOpts {
  /** Top-K cap on returned tiles. Default 2 — matches Python. */
  limit?: number;
  /** Threshold below which a tile is excluded entirely. Default 3.5. */
  minHeat?: number;
}

/**
 * Top-K hottest tiles, ordered desc. Mirrors `hot_tiles`.
 * Tie-breaking by tileId asc (Python sorted() is stable; insertion
 * order across `world.all_tiles()` is the upstream contract — we
 * accept the same convention).
 */
export function hotTiles(
  tiles: ReadonlyArray<TileForHeat>,
  opts: HeatOpts = {},
): TileForHeat[] {
  const limit = opts.limit ?? 2;
  const minHeat = opts.minHeat ?? 3.5;
  const scored: Array<{ tile: TileForHeat; heat: number }> = [];
  for (const t of tiles) {
    const h = scoreTileHeat(t);
    if (h >= minHeat) scored.push({ tile: t, heat: h });
  }
  // sort desc by heat; stable so input order persists for ties.
  scored.sort((a, b) => b.heat - a.heat);
  return scored.slice(0, limit).map((s) => s.tile);
}
